opengrok-indexer/src/main/java/org/opengrok/indexer/util/Progress.java

```diff
@@ -0,0 +1,142 @@
+ * Portions Copyright (c) 2017-2018, Chris Fraire <cfraire@me.com>.
+ */
+
+package org.opengrok.indexer.util;
+
+import org.opengrok.indexer.configuration.RuntimeEnvironment;
+
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class Progress implements AutoCloseable {
+    private final Logger logger;
+    private final long totalCount;
+    private final String suffix;
+
+    private AtomicLong currentCount = new AtomicLong();
+    private Thread loggerThread = null;
+    private volatile boolean run;
+
+    private final Object sync = new Object();
+
+    /**
+     * @param logger logger instance
+     * @param suffix string suffix to identify the operation
+     * @param totalCount total count
+     */
+    public Progress(Logger logger, String suffix, long totalCount) {
+        this.logger = logger;
+        this.suffix = suffix;
+        this.totalCount = totalCount;
+
+        // Assuming printProgress configuration setting cannot be changed on the fly.
+        if (totalCount > 0 && RuntimeEnvironment.getInstance().isPrintProgress()) {
+            // spawn a logger thread.
+            run = true;
+            loggerThread = new Thread(this::logLoop,
+                    "progress-thread-" + suffix.replaceAll(" ", "_"));
+            loggerThread.start();
+        }
+    }
+
+    /**
+     * Increment counter. The actual logging will be done eventually.
+     */
+    public void increment() {
+        this.currentCount.incrementAndGet();
+
+        if (loggerThread != null) {
+            // nag the thread.
+            synchronized (sync) {
+                sync.notify();
+            }
+        }
+    }
+
+    private void logLoop() {
+        long cachedCount = 0;
+
+        while (true) {
+            if (!run) {
+                return;
+            }
+
+            long currentCount = this.currentCount.get();
+            Level currentLevel;
+
+            // Do not log if there was no progress.
+            if (cachedCount >= currentCount) {
+                continue;
+            }
+
+            if (currentCount <= 1 || currentCount % 100 == 0) {
+                currentLevel = Level.INFO;
+            } else if (currentCount % 50 == 0) {
+                currentLevel = Level.FINE;
+            } else if (currentCount % 10 == 0) {
+                currentLevel = Level.FINER;
+            } else {
+                currentLevel = Level.FINEST;
+            }
+
+            if (logger.isLoggable(currentLevel)) {
+                logger.log(currentLevel, "Progress: {0} ({1}%) for {2}",
+                        new Object[]{currentCount, currentCount * 100.0f /
+                                totalCount, suffix});
+            }
+
+            cachedCount = currentCount;
+
+            // wait for event
+            try {
+                synchronized (sync) {
+                    if (!run) {
+                        return;
```

Comment:
>> Do you think you might accidentally forgo logging the final entry by returning here?

>It's actually deadlock prevention. If the thread calling close() notifies the loop thread while it is still executing the loop body (before the wait() call) they will wait for each other indefinitely.

I meant maybe it's better to `continue` instead of `return` to be sure to do the final logging. I recommend moving the `run` check of line 82 right before this try block — i.e. check `run` right after logging and in an unsynchronized context. Still keep this `synchronized` check, and it's a nice double-checked locking pattern.

Reply:
I see, thanks.